Cookbook/Proposals/FeatureSwitchesRedesign.md

```diff
@@ -0,0 +1,218 @@
+
+To solve these issues we propose to replace `protocol AppConfigurationProtocol` with concrete `struct AppConfiguration` defined in `BabylonDependencies` and initialise it with application specific values in a target specific code. This will reduce amount of boilerplate to write, but will still require us to change code multiple times in each target when something changes in this struct. I.e. when we add a new feature switch to `AppConfiguration` we will need to add it to the paramters that we pass to it's initializer in each target. We can use default values where possible to avoid that. For that we should agree that default values are always `false` so that we can definetely know what is the value if it is not passed to the constructor. Protocol and default protocol implementations achive the same result but with much more boilerplate (remember point-free issue about using value types instead of protocols)
+
+All feature specific configuration should be moved to the corresponding feature frameworks and should use concrete structs instead of protocols that then `AppConfiguration` extends. `AppConfiguration` should only be concerned with application level configurations (i.e. enabled tabs or privacy notices urls).
+
+3. All feature switches access points should be moved to the `Current` and consequentially should be only used in builders - view models, flow controllers and renderers should have them injected through their constructors. Currently we only have `Current.abTestingService` but we have global `LocalFeatureSwitches`
+
+4. Root.plist should be generated based on the code that declared feature switches.
+
+For that we can use SwiftSyntax to analyze the content of the files which define feature switches (based on the file naming convetion) and generate entry in the plist for each of them. This way we don't have to deal with plist manually and have a single source of thruth for defining feature switches. The plist will be ignored by git and will be generated as part of build step only in DEBUG configuration, so we don't have to strip it from release builds.
+
+
+## Impact on existing codebase
+
+Currently used types for feature switches should be deprecated and either replaced with a new approach in one go or gradually. We can keep `ABTestingService` related types but repurpose them to serve as providers of the feature switches rather than 
+
+## Alternatives considered
+
+- Keep all the feature switches in one place in the `BabylonDependencies`. The main downside of this approach if that it goes against modularity - we will need to change flags in the `BabylonDependencies` and rebuild it instead of changing it just in the feature framework and rebuilding only it. It can also trigger rebuild of all other frameworks depending on the `BabylonDependencies` as we know how Xcode incremental builds are unrelyiable.
```

Comment:
```suggestion
- Keep all the feature switches in one place in the `BabylonDependencies`. The main downside of this approach if that it goes against modularity - we will need to change flags in the `BabylonDependencies` and rebuild it instead of changing it just in the feature framework and rebuilding only it. It can also trigger rebuild of all other frameworks depending on the `BabylonDependencies` as we know how Xcode incremental builds are unreliable.
```